service/s3/handwritten-paginators.go

```diff
@@ -0,0 +1,193 @@
+package s3
+
+import (
+	"context"
+	"fmt"
+)
+
+// ListObjectVersionsAPIClient is a client that implements the ListObjectVersions
+// operation
+type ListObjectVersionsAPIClient interface {
+	ListObjectVersions(context.Context, *ListObjectVersionsInput, ...func(*Options)) (*ListObjectVersionsOutput, error)
+}
+
+var _ ListObjectVersionsAPIClient = (*Client)(nil)
+
+// ListObjectVersionsPaginatorOptions is the paginator options for ListObjectVersions
+type ListObjectVersionsPaginatorOptions struct {
+	// (Optional) The maximum number of Object Versions that you want Amazon Route 53 to
+	// return.
+	Limit int32
+
+	// Set to true if pagination should stop if the service returns a pagination token
+	// that matches the most recent token provided to the service.
+	StopOnDuplicateToken bool
+}
+
+// ListObjectVersionsPaginator is a paginator for ListObjectVersions
+type ListObjectVersionsPaginator struct {
+	options         ListObjectVersionsPaginatorOptions
+	client          ListObjectVersionsAPIClient
+	params          *ListObjectVersionsInput
+	firstPage       bool
+	KeyMarker       *string
+	VersionIdMarker *string
+}
+
+// NewListObjectVersionsPaginator returns a new ListObjectVersionsPaginator
+func NewListObjectVersionsPaginator(client ListObjectVersionsAPIClient, params *ListObjectVersionsInput, optFns ...func(*ListObjectVersionsPaginatorOptions)) *ListObjectVersionsPaginator {
+	if params == nil {
+		params = &ListObjectVersionsInput{}
+	}
+
+	options := ListObjectVersionsPaginatorOptions{}
+	options.Limit = params.MaxKeys
+
+	for _, fn := range optFns {
+		fn(&options)
+	}
+
+	return &ListObjectVersionsPaginator{
+		options:         options,
+		client:          client,
+		params:          params,
+		firstPage:       true,
+		KeyMarker:       params.KeyMarker,
+		VersionIdMarker: params.VersionIdMarker,
+	}
+}
+
+// HasMorePages returns a boolean indicating whether more pages are available
+func (p *ListObjectVersionsPaginator) HasMorePages() bool {
+	return p.firstPage || (p.KeyMarker != nil && len(*p.KeyMarker) != 0)
+}
+
+// NextPage retrieves the next ListObjectVersions page.
+func (p *ListObjectVersionsPaginator) NextPage(ctx context.Context, optFns ...func(*Options)) (*ListObjectVersionsOutput, error) {
+	if !p.HasMorePages() {
+		return nil, fmt.Errorf("no more pages available")
+	}
+
+	params := *p.params
+	params.KeyMarker = p.KeyMarker
+	params.VersionIdMarker = p.VersionIdMarker
+
+	var limit int32
+	if p.options.Limit > 0 {
+		limit = p.options.Limit
+	}
+	params.MaxKeys = limit
+
+	result, err := p.client.ListObjectVersions(ctx, &params, optFns...)
+	if err != nil {
+		return nil, err
+	}
+	p.firstPage = false
+
+	prevToken := p.KeyMarker
+
+	p.KeyMarker = result.NextKeyMarker
+	p.VersionIdMarker = result.NextVersionIdMarker
+
+	if p.options.StopOnDuplicateToken &&
```

Comment:
From the S3 docs:
>Only Amazon S3 generates version IDs, and they cannot be edited. Version IDs are Unicode, UTF-8 encoded, URL-ready, opaque strings that are no more than 1,024 bytes long. The following is an example:
>
>3sL4kqtJlcpXroDTDmJ+rmSpXd3dIbrHY+MTRCxf3vjVBH40Nr8X8gdRQBpUMLUo

Since they are unique identifiers, and cannot be set by the user, the chance of collision within a specific bucket are tiny. 
What are your thoughts?

Reply:
I would think it's more about any one parameter being a duplicate rather than collision being the concern. I'm not 100% sure though, maybe another thing to look to other SDKs w.r.t behavior.